### PR TITLE
feat: add config to control write all status to x-upsream-apisix-status

### PR DESCRIPTION
**Pull request by liangliang4ward**
### What this PR does / why we need it:

config to control return all status at `X-APISIX-Upstream-Status`

https://github.com/apache/apisix/issues/3963
### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/init.lua [hunk 1]
```diff
@@ -589,25 +589,33 @@ function _M.http_header_filter_phase()
...
-        if tonumber(last_status) >= 500 and tonumber(last_status) <= 599 then
-            set_resp_upstream_status(up_status)
+                if tonumber(last_status) >= 500 and tonumber(last_status) <= 599 then
+                    set_resp_upstream_status(up_status)
+                end
+            end
         end
     end
 
+
```
Comment (tzssangglass) [inline apisix/init.lua:618]: delete this blank line

File: docs/zh/latest/debug-function.md [hunk 2]
```diff
@@ -23,7 +23,7 @@ title: 调试功能
 
 ## `5xx` 响应状态码
 
-500、502、503 等类似的 `5xx` 状态码，是由于服务器错误而响应的状态码，当一个请求出现 `5xx` 状态码时；它可能来源于 `APISIX` 或 `Upstream` 。如何识别这些响应状态码的来源，是一件很有意义的事，它能够快速的帮助我们确定问题的所在。
+500、502、503 等类似的 `5xx` 状态码，是由于服务器错误而响应的状态码，当一个请求出现 `5xx` 状态码时；它可能来源于 `APISIX` 或 `Upstream` 。如何识别这些响应状态码的来源，是一件很有意义的事，它能够快速的帮助我们确定问题的所在。(当修改 `conf/nginx.conf` 的配置 `enable_all_upstream_status` 为 true 时，会将上游所有状态都返回，不仅仅是5XX状态。)
```
Comment (tzssangglass) [inline docs/zh/latest/debug-function.md:26]:
```suggestion
500、502、503 等类似的 `5xx` 状态码，是由于服务器错误而响应的状态码，当一个请求出现 `5xx` 状态码时；它可能来源于 `APISIX` 或 `Upstream` 。如何识别这些响应状态码的来源，是一件很有意义的事，它能够快速的帮助我们确定问题的所在。(当修改 `conf/nginx.conf` 的配置 `enable_all_upstream_status` 为 true 时，会返回所有上游状态码，不仅仅是 `5xx` 状态码。)
```

File: t/node/upstream-status-all.t [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+#
+use t::APISIX 'no_plan';
+
+repeat_each(1);
+log_level('info');
+worker_connections(256);
+no_root_location();
+no_shuffle();
+
+run_tests();
```
Comment (leslie-tsang) [inline t/node/upstream-status-all.t:25]:
```suggestion
no_shuffle();

add_block_preprocessor(sub {
    my ($block) = @_;

    if (!$block->request) {
        $block->set_value("request", "GET /t");
    }

    if ((!defined $block->error_log) && (!defined $block->no_error_log)) {
        $block->set_value("no_error_log", "[error]");
    }
});

run_tests();
```
Would be better reduce duplicate sections in `add_block_preprocessor`

File: t/node/upstream-status-all.t [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+                )
+
+            if code >= 300 then
+                ngx.status = code
+            end
+            ngx.say(body)
+        }
+    }
+--- request
+GET /t
```
Comment (leslie-tsang) [inline t/node/upstream-status-all.t:59]:
```suggestion
```
Need to be removed.

File: t/node/upstream-status-all.t [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+--- response_body
+passed
+--- no_error_log
+[error]
```
Comment (leslie-tsang) [inline t/node/upstream-status-all.t:63]:
```suggestion
```
Ditto

**Review by leslie-tsang — CHANGES_REQUESTED** [review]

File: t/node/upstream-status-all.t (hunk 5 above)
Comment (liangliang4ward) [reply t/node/upstream-status-all.t]: done

File: t/node/upstream-status-all.t (hunk 4 above)
Comment (liangliang4ward) [reply t/node/upstream-status-all.t]: done

File: t/node/upstream-status-all.t (hunk 3 above)
Comment (liangliang4ward) [reply t/node/upstream-status-all.t]: ok

File: apisix/init.lua (hunk 1 above)
Comment (liangliang4ward) [reply apisix/init.lua]: ok

File: apisix/init.lua [hunk 6]
```diff
@@ -589,22 +589,29 @@ function _M.http_header_filter_phase()
...
-       and tonumber(up_status) <= 599
-    then
-        set_resp_upstream_status(up_status)
-    elseif up_status and #up_status > 3 then
-        -- the up_status can be "502, 502" or "502, 502 : "
-        local last_status
-        if str_byte(up_status, -1) == str_byte(" ") then
-            last_status = str_sub(up_status, -6, -3)
+    if up_status then
+        local_conf = core.config.local_conf();
```
Comment (tokers) [inline apisix/init.lua:593]:
```suggestion
        local_conf = core.config.local_conf()
```

File: apisix/init.lua [hunk 7]
```diff
@@ -589,22 +589,29 @@ function _M.http_header_filter_phase()
...
+        if local_conf.apisix and local_conf.apisix.enable_all_upstream_status then
```
Comment (tokers) [inline apisix/init.lua:594]: The setting name is not so intuitive, what about `show_upstream_status_in_response_header`?

File: docs/zh/latest/config.json [hunk 8]
```diff
@@ -116,9 +116,8 @@
               "type": "category",
               "label": "Metrics",
               "items": [
-                "plugins/promethues",
-                "plugins/node-status",
-                "plugins/datadog"
```
Comment (tokers) [inline docs/zh/latest/config.json:121]: Why remove it?

File: apisix/init.lua (hunk 7 above)
Comment (liangliang4ward) [reply apisix/init.lua]: That's a good idear

Comment (liangliang4ward) [reply apisix/init.lua]: image

File: docs/zh/latest/config.json (hunk 8 above)
Comment (liangliang4ward) [reply docs/zh/latest/config.json]: image

Comment (tokers) [reply docs/zh/latest/config.json:121]: It seems that we should add the Chinese doc, I have submitted an issue to track this: https://github.com/apache/apisix/issues/6406.

Comment (liangliang4ward) [reply docs/zh/latest/config.json:121]: yes,Yes. But it shouldn't be done in this PR.

Comment (liangliang4ward) [reply docs/zh/latest/config.json]: master branch already fix .

File: t/node/upstream-status-all.t [hunk 9]
```diff
@@ -0,0 +1,461 @@
...
+=== TEST 2: hit the route and $upstream_status is 200
+--- yaml_config
+apisix:
+  show_upstream_status_in_response_header: true
+--- request
+GET /hello
+--- response_body
+hello world
+--- error_log
+X-APISIX-Upstream-Status: 200
```
Comment (tokers) [inline t/node/upstream-status-all.t:79]: Since this is a response header, why we need to verify it by error log entry?

Comment (liangliang4ward) [reply t/node/upstream-status-all.t:79]: maybe where response status =200 , do not record log? 
The test case use the request header to judge whether it is successful?
What's your opinion?

**Review by tzssangglass — DISMISSED** [review]

File: t/node/upstream-status-all.t (hunk 9 above)
Comment (liangliang4ward) [reply t/node/upstream-status-all.t:76]: fixed

**Review by tzssangglass — APPROVED** [review]

**Review by tokers — APPROVED** [review]

**Review by spacewander — APPROVED** [review]

**Pull request merged**